<h1 align = center> Neighborhoods in Toronto </h1>
    
    
    
<p>here we will scrap the data from wekipidea page using BeautifulSoup library and than convert this data into pandas dataframe    

In [1]:
#import the library
import numpy as np
import pandas as pd
#Library used to pulling data from html file
from bs4 import BeautifulSoup
# library used for fatching internet resources
from urllib.request import urlopen

In [2]:
# url of the wikipedia page
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = urlopen(url)

In [3]:
soup = BeautifulSoup(html , 'lxml')
type(soup)

bs4.BeautifulSoup

In [4]:
#title of the web page
title = soup.title.text
title

'List of postal codes of Canada: M - Wikipedia'

In [5]:
# content of the web page
content = soup.find(id = 'content')
text_content = content.find(id = 'bodyContent')

In [6]:
#tabular content of the web page
table = text_content.find('table')
table_content = table.tbody.find_all('tr')
len(table_content)

181

In [7]:
#find the heading of the table
heading = []
for head in table_content[0].find_all('th'):
    heading.append(head.text.replace('\n','').strip())
heading

['Postal Code', 'Borough', 'Neighborhood']

In [8]:
table_content = table_content[1:]
table_content[1].find_all('td')
len(table_content)

180

In [9]:
#extract the tabular data from web page
PostalCode  = []
Borough = []
Neighborhood = []
for data in table_content:
    data = data.find_all('td')
    PostalCode.append(data[0].text.replace('\n', '').strip())
    Borough.append(data[1].text.replace('\n', '').strip())
    Neighborhood.append(data[2].text.replace('\n','').strip())
  

In [10]:
#convert the tabular data into the dictionary
data = {'PostalCode': PostalCode,
       'Borough': Borough,
       'Neighborhood': Neighborhood}

In [11]:
#convert this dictionary into dataframe
neighborhood_data = pd.DataFrame(data)
neighborhood_data.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

In [12]:
#shape of the dataframe
neighborhood_data.shape

(180, 3)

In the `neighborhood_data` we have a 180 rows and 3 columns

**Drop all the rows that have Borough is not assigned**

In [13]:
neighborhood_data = neighborhood_data[neighborhood_data.Borough != 'Not assigned']
neighborhood_data.shape

(103, 3)

After droping all the rows that have Borough is not assigned, now we have 103 rows and 3 columns

In [14]:
neighborhood_data[neighborhood_data.PostalCode =='M5A']

PostalCode           Borough               Neighborhood
4        M5A  Downtown Toronto  Regent Park, Harbourfront

In [15]:
neighborhood_data.head(10)

PostalCode           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge
11        M3B        North York                                    Don Mills
12        M4B         East York              Parkview Hill, Woodbine Gardens
13        M5B  Downtown Toronto                     Garden District, Ryerson

**if borough isassigned and Neighborhood is not assigned than assign the borough to the neighborhood**

In [16]:
neighborhood_data[neighborhood_data.Neighborhood == 'Not assigned']

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

There is no  row available there Boriough is assigned and Neighborhood is not assigned

In [17]:
# the shape of the clean data
print('clean data has a {} rows and {} columns'.format(neighborhood_data.shape[0], neighborhood_data.shape[1]))

clean data has a 103 rows and 3 columns


In [19]:
#save this data into csv file for next time use
neighborhood_data.to_csv('neighborhood_data.csv', index = False)